[keyboard] support coding input method for chinese

taxigps opened the pull request:
There are thousands of characters in chinese, so can't input all the characters driect from keys on keyboard. the method is use ascii keys to input a coding string, and search the code in a coding table to get the chinese character or words want to input. one code will return many results, so show them in a list to choose.

the BaiduPY coding table is a cloud service from Baidu.

a sample screenshot here:
![screenshot087_1](https://cloud.githubusercontent.com/assets/102434/4178245/dfcf7a42-3686-11e4-9a3f-f2ac4459c7b3.jpg)

MartijnKaijser commented:
@mkortstiege maybe you could take a look to get this possibly in

mkortstiege commented:
I think that´s something for @Paxxi

Paxxi commented:
@MartijnKaijser @mkortstiege I don't currently know much about Chinese but I'll try and get this sorted.

Paxxi commented:
@taxigps can you rebase this on master?

Is this pinyin or is this some other method? The baidu pages have been a bit tricky

taxigps commented:
@Paxxi rebased. Yes, It's a kind of pinyin input method.

Paxxi commented:
generally it looks okay but I think it needs a bit more work until ready.
1. You should try to use JSONVariantParser and parse the json the right way and not with regexps
2. The api call to baidu should be on a separate thread, blocking the ui thread is not a good thing and it makes for a bad experience.
On to the smaller things
* CInputCodingTable should be called IInputCodingTable since it's an interface
* you use (WCHAR) as a cast in several places, please use wchar_t instead
* It would be nice for touch users if the suggestions were buttons and clickable(this is a secondary goal though and maybe not feasible)

ping me if there's any questions or if you need any help

Montellese commented:
Haven't looked at the changes to the keyboard dialog yet and I also don't really know how this system works but the current GUI integration looks very hacked in and I'm wondering if there'd be a better way to integrate this. @ronie @BigNoid any suggestions?

taxigps commented:
@Paxxi how to use JSONVariantParser? I think the data is not strict json data. like this:
```
[[["\u554a",1,{"pinyin":"a","type":"IMEDICT"}],["\u963f\u0073\u0075\u0065",1,{"pinyin":"a","type":"NEWWORD"}],["\u963f\u0062\u0065\u006e",1,{"pinyin":"a","type":"NEWWORD"}],["\u963f",1,{"pinyin":"a","type":"IMEDICT"}],["\u5416",1,{"pinyin":"a","type":"IMEDICT"}],["\u814c",1,{"pinyin":"yan","type":"IMEDICT"}],["\u55c4",1,{"pinyin":"a","type":"IMEDICT"}],["\u9515",1,{"pinyin":"a","type":"IMEDICT"}],["\u5475",1,{"pinyin":"a","type":"IMEDICT"}]],"a"]
```

Montellese commented:
@taxigps: Isn't `\u554a` a unicode character? In that case you should probably use `itemdata[0].asWideString()`?

taxigps commented:
@Montellese have tried `asWideString()`, can't retrun right string too.

taxigps commented:
@Montellese added a local codingtable for people without internet access.
@Paxxi added a separate thread for baidu api call.
and JSONVariantParser can't work.

Paxxi commented:
@taxigps I threw together one way to do it in a background thread. It's not finished unfortunatley because I don't have any more time today. It should give you an idea of how to accomplish it though.
you can find it here https://github.com/Paxxi/xbmc/tree/chinese

Paxxi commented:
the local table you added got me googling again how large a dataset would be to skip the web api alltogheter and led me to this https://github.com/libpinyin/libpinyin . Haven't had time to look in depth at it but maybe it's a better idea? I'll try and get some time tomorrow to go over it.

taxigps commented:
@Paxxi If we can use libpinyin to support chinese input is best. My method is simple, but not good. my local codingtable just support input chinese charaters, not support word and sentence input.

Paxxi commented:
Sorry about the slow response, I've gotten sidetracked a lot. Libpinyin is not an option, it depends on gtk and thats just not something we can use. Will get back to you when I've gone through your latest changes.

ronie commented:
i would add a separate section on top (or below) the virtual keyboard.
something like this:
![pr5326](https://cloud.githubusercontent.com/assets/687265/6761580/1293cf34-cf55-11e4-909f-6d665f5b505e.jpg)

code: https://github.com/ronie/xbmc/commit/f6dc348863745ff37ab5094ef69e3b5c7de62b5c

Montellese commented:
@ronie: Yeah I thought about something similar but the approach with pressing 0, 1, 2, etc. is not gonna work on any installation that is not using a keyboard or a remote with a lot of buttons. IMO this should be a list (maybe on the side?) where the user can navigate to or touch into to select something.

ronie commented:
if @taxigps can code it that way, i don't mind handling the skinning work.

taxigps commented:
@ronie skinners can put this two controls any where as they like. I'll merge your code. :)  
@Montellese may be change HZLIST label to ten button so people can touch to select? but I think it's the same work between touch 0-9 on the virtual keyboard then touch the ten chinese words.

Montellese commented:
Ah the virtual 0-9 buttons work for this as well? I assumed they would result in the respective number being written to the input text. What I don't like is the limit of 10 values. With an additional list view you could show a lot more buttons by letting the users scroll the list. But I have no idea if that is feasible or not (i.e. it depends on the usual length of the list).

taxigps commented:
@Paxxi find an other source code for pinyin input, and want to get this in. https://github.com/taxigps/freepy

taxigps commented:
@Montellese A good chinese input method tend to high efficiency. So open an other list is not good choice. A good coding table not have many words in the same code, and let the most frequently used words at first ten place. This why some input method open source for their code but don't open source for their coding table. like this one: https://github.com/thunisoft/unispim

Montellese commented:
@taxigps: I didn't mean an extra dialog but instead of a bunch of labels which are limited to 10 items I would propose to use a list control which can be scrolled. You could still use the 0-9 keys as shortcuts for the first 10 symbols to optimize it for keyboard input.
But since I don't really know how these keyboards work I'm leaving the decision up to you and the skinners.

Paxxi commented:
@taxigps that library seems to be windows only so best method is still your first web service. I'm not sure about Chinese but how often does anyone scroll through the autocomplete list on a phone to find a word? 10 items should be fine and in line with how others does it.

taxigps commented:
rebased and changed skin code to https://github.com/ronie/xbmc/commit/f6dc348863745ff37ab5094ef69e3b5c7de62b5c
and still thinking better way to use thread like https://github.com/Paxxi/xbmc/tree/chinese

taxigps commented:
@Paxxi changed to use thread.

taxigps committed:
[keyboard] support codingtable input and add Chinese BaiduPY/BasePY layouts

taxigps commented:
jenkins build this please

taxigps commented:
jenkins build this please

taxigps commented:
jenkins build this please

taxigps commented on xbmc/input/InputCodingTable.h:
```diff
@@ -0,0 +1,35 @@
+public:
+  virtual ~IInputCodingTable() {}
+  virtual bool GetWordListPage(const std::string& strCode, bool isFirstPage) = 0;
+  virtual std::shared_ptr<std::vector<std::wstring>> GetResponse(int response) = 0;
```
@Paxxi this line works well under windows, but cause error under linux:
`error: 'shared_ptr' in namespace 'std' does not name a type`
how to solve this?

Paxxi replied:
No need to use a shared_ptr, just return the vector

taxigps commented on xbmc/input/InputCodingTableBaiduPY.h:
```diff
@@ -0,0 +1,52 @@
+  int m_messageCounter;
+
+  std::deque<std::string> m_work;
+  std::map<int, std::shared_ptr<std::vector<std::wstring>>> m_responses;
```
@Paxxi this line cause error under linux and osx, like:
`error: no template named 'map' in namespace 'std'`
how to solve this?

Paxxi replied:
Add #include <map>

Paxxi replied:
Also no need for the shared_ptr here either

taxigps commented:
jenkins build this please

taxigps commented:
@Paxxi still need your help. now build success under wn32, linux and osx. but still failure under android. I don't know what's the problem now.

Paxxi commented:
I'm not really sure about the failure on android and why pcre includes would behave like that. @Memphiz or @koying can you take a look at the build failure?

koying commented:
Try including <stdlib.h> as first include of xbmc/input/InputCodingTableBaiduPY.cpp

It's an android peculiarity with stdlib

koying commented:
Silly markup... it's stdlib.h that you have to include ;)

taxigps commented:
@Paxxi @koying thanks!
jenkins build this please

taxigps commented:
@koying include stdlib.h but still build failure.

koying commented:
Mmmm... I'll try to build it at home

koying commented:
Meh, builds for me. A branch of mine just with this also builds on jenkins???
jenkins build this, please

koying commented:
jenkins build this please

koying commented:
Yep, latest jenkins built. Might be a jenkins issue...

taxigps commented:
@koying thanks！
@Paxxi is there any other thing need to do?

Paxxi commented on xbmc/input/InputCodingTableBaiduPY.cpp:
```diff
@@ -0,0 +1,140 @@
+  if (strCode != m_code) // don't handle obsolete response
+    return;
+
+  auto words = std::make_shared<std::vector<std::wstring>>();
```
get rid of the shared_ptr, just use a local std::vector since it'll get copied/moved into the map before we exit the method

Paxxi commented on xbmc/input/InputCodingTableBaiduPY.cpp:
```diff
@@ -0,0 +1,140 @@
+
+bool CInputCodingTableBaiduPY::GetWordListPage(const std::string& strCode, bool isFirstPage)
+{
+  if (!strCode.length())
```
change to if (strCode.empty())

Paxxi commented on xbmc/input/InputCodingTable.h:
```diff
@@ -0,0 +1,36 @@
+
+#include <string>
+#include <vector>
+#include <memory>
```
< memory > can be removed now, it's not used

Paxxi commented on xbmc/input/InputCodingTableBasePY.h:
```diff
@@ -0,0 +1,451 @@
+  virtual ~CInputCodingTableBasePY() {}
+
+  virtual bool GetWordListPage(const std::string& strCode, bool isFirstPage);
+  virtual std::vector<std::wstring> GetResponse(int response);
```
you can remove response here so it's GetResponse(int) to avoid warnings about unused variables, same for .cpp file

Paxxi commented on xbmc/input/InputCodingTableBasePY.h:
```diff
@@ -0,0 +1,451 @@
+*/
+
+#include "InputCodingTable.h"
+#include <map>
```
shoulde include < string > as well

Paxxi commented on xbmc/input/InputCodingTableBasePY.cpp:
```diff
@@ -0,0 +1,57 @@
+    return false;
+
+  m_words.clear();
+  std::map<std::string, std::wstring>::const_iterator finder = m_mapHZCode.find(strCode);
```
you can use auto here to avoid some typing

Paxxi commented on xbmc/input/KeyboardLayout.h:
```diff
@@ -56,4 +58,5 @@ class CKeyboardLayout
   std::string m_language;
   std::string m_layout;
   Keyboards m_keyboards;
+  IInputCodingTable *m_codingtable;
```
make this a std::unique_ptr since it's the only one owning it, the pointer in GUIDialogKeyboardGeneric can still be a plain pointer.

taxigps replied:
so change this line to
```
std::unique_ptr<IInputCodingTable> m_codingtable;
```
and line 34 to:
```
  IInputCodingTable* GetCodingTable() { return m_codingtable.get(); }
```
this change cause compile error in `CGUIDialogKeyboardGeneric::UpdateButtons`:
```
error C2280: 'std::unique_ptr<IInputCodingTable,std::default_delete<_Ty>>::unique_ptr(const std::unique_ptr<_Ty,std::default_delete<_Ty>> &)' : attempting to reference a deleted function
```

taxigps replied:
@Paxxi any advice?

Paxxi commented:
There's still some nitpicking like c-style casts left but I think this is in good enough shape to get merged. Nice work @taxigps overall.
@MartijnKaijser @Montellese what's your oppinions about this? good to go except for the things I pointed out?

taxigps commented:
@Paxxi rebased and update codes as your advice.  but not change m_codingtable to std::unique_ptr because some error.

Paxxi commented:
I think this should be good to merge after a rebase. Any objections @Montellese @ronie @mkortstiege ?

ronie commented:
i fine with it.
@phil65 mind having a look at it as well?

phil65 commented on addons/skin.confluence/720p/DialogKeyboard.xml:
```diff
@@ -20,6 +20,54 @@
+			<control type="group">
+				<left>0</left>
+				<top>-90</top>
+				<visible>Control.IsVisible(313)</visible>
```
Being nitpicky: I dont really like group controls which toggle visibility based on a control which is part of the group itself. It would be more clear if 313 would be outside of this group.

ronie replied:
agreed. it would be better if we had an infobool we could check instead.
something like Keyboard.HasCodingInput or similar.

taxigps replied:
how to use infobool?

phil65 commented:
One small suggestion (but more "cosmetics"), apart from that skin code looks fine.

Montellese commented on xbmc/dialogs/GUIDialogKeyboardGeneric.cpp:
```diff
@@ -129,7 +176,9 @@ void CGUIDialogKeyboardGeneric::OnInitWindow()
 {
   bool handled = true;
-  if (action.GetID() == ACTION_ENTER || (m_isKeyboardNavigationMode && action.GetID() == ACTION_SELECT_ITEM))
+  if (action.GetID() == (KEY_VKEY | 8))
```
Is there a define for what `8` means? Otherwise it's hard to understand what it means.

Montellese commented on xbmc/dialogs/GUIDialogKeyboardGeneric.cpp:
```diff
@@ -143,26 +192,33 @@ bool CGUIDialogKeyboardGeneric::OnAction(const CAction &action)
-    if (edit)
-      handled = edit->OnAction(action);
-    if (!handled && action.GetID() >= KEY_VKEY && action.GetID() < KEY_ASCII)
+    std::wstring wch=L"";
```
Spaces around the `=`

Montellese commented on xbmc/dialogs/GUIDialogKeyboardGeneric.cpp:
```diff
@@ -257,6 +313,17 @@ bool CGUIDialogKeyboardGeneric::OnMessage(CGUIMessage& message)
+  case GUI_CODINGTABLE_LOOKUP_COMPLETED:
+    {
+      int response = message.GetParam1();
+      auto words = m_codingtable->GetResponse(response);
```
This is only used if the condition below is true so this can be moved inside the conditional block.

Montellese commented on xbmc/dialogs/GUIDialogKeyboardGeneric.cpp:
```diff
@@ -295,10 +367,19 @@ void CGUIDialogKeyboardGeneric::Character(const std::string &ch)
-  CGUIControl *edit = GetControl(CTL_EDIT);
-  if (edit)
-    edit->OnAction(CAction(ACTION_BACKSPACE));
+  if (m_codingtable && m_hzcode.length()>0)
```
Spaces around the `>`

Montellese commented on xbmc/dialogs/GUIDialogKeyboardGeneric.h:
```diff
@@ -64,6 +66,11 @@ class CGUIDialogKeyboardGeneric : public CGUIDialog, public CGUIKeyboard
     void Backspace();
     void SetEditText(const std::string& text);
     void SendSearchMessage();
+    float getStringWidth(const std::wstring& utf16);
```
`GetStringWidth`

Montellese commented on xbmc/dialogs/GUIDialogKeyboardGeneric.cpp:
```diff
@@ -478,3 +575,115 @@ bool CGUIDialogKeyboardGeneric::ShowAndGetInput(char_callback_t pCallback, const
+    ChangeWordList(0);
+    return true;
+  }
+  else if (ch[0] >= '0' && ch[0] <= '9')
```
No need for `else`

Montellese commented on xbmc/dialogs/GUIDialogKeyboardGeneric.h:
```diff
@@ -64,6 +66,11 @@ class CGUIDialogKeyboardGeneric : public CGUIDialog, public CGUIKeyboard
     void SendSearchMessage();
+    float getStringWidth(const std::wstring& utf16);
+    void ChangeWordList(int direct);  // direct: 0 - first page, 1 - next page, -1 - prev page
+    void _ChangeWordList(int direct); // direct: 0 - current page, 1 - next page, -1 -prev page
```
Please find a better name for this method as there's no way to know the difference when seeing the method names.

Montellese commented on xbmc/input/InputCodingTable.h:
```diff
@@ -0,0 +1,35 @@
+#include <string>
+#include <vector>
+
+#define GUI_CODINGTABLE_LOOKUP_COMPLETED 65000 //pick something high to avoid conflicts with other gui messages
```
Any reason why this isn't defined where all the other user GUI messages are defined?

Paxxi replied:
This one's my fault. I stuck it here since it's not of any interest to other windows but maybe it's better to have it with the other?

Montellese replied:
IMO all "custom" GUI_MSG_FOO defines should be in `GUIUserMessages.h` to avoid any conflicts which are much easier to see if all the definitions are in the same file.
And maybe also rename it to `GUI_MSG_CODINGTABLE_LOOKUP_COMPLETED` to match the naming of the rest.

Montellese commented on xbmc/input/InputCodingTable.h:
```diff
@@ -0,0 +1,35 @@
+  virtual ~IInputCodingTable() {}
+  virtual bool GetWordListPage(const std::string& strCode, bool isFirstPage) = 0;
+  virtual std::vector<std::wstring> GetResponse(int response) = 0;
+  std::string m_codechars;
```
Shouldn't this be protected/private and have methods to access whatever we need from it?

Montellese commented on xbmc/input/InputCodingTableBaiduPY.cpp:
```diff
@@ -0,0 +1,140 @@
+      strUrl = StringUtils::Format("http://olime.baidu.com/py?input=%s&inputtype=py&bg=%d&ed=%d&result=hanzi&resultcoding=unicode&ch_en=0&clientinfo=web", work.c_str(), m_api_begin, m_api_end);
+
+      if (http.Get(strUrl, data))
+      {
```
No need for the curly braces.

Montellese commented on xbmc/input/InputCodingTableBasePY.h:
```diff
@@ -0,0 +1,452 @@
+private:
+  std::vector<std::wstring> m_words;
+  static const std::map<std::string, std::wstring> m_mapHZCode;
+  static std::map<std::string, std::wstring> CreateHZCodeMap()
```
Should this implementation really be in the header? Furthermore C++11 supports map initialization lists so we don't need the static method workaround anymore.

Montellese commented:
I don't remember if I already mentioned these concerns in the past argument so I'll just write them down again:
* I really don't like the hardcoded URL to the online service in the C++ code. If that URL ever changes (for whatever reasons) we can't fix it without recompiling Kodi and making a new release.
* I think it would be nicer if these coding table implementations could be done with (python) addons to make it easier to also extend it for other languages.

Apart from that there are only minor code changes (for which I added comments) necessary.

Paxxi commented:
Thanks @Montellese. Do you have an idea to move the url out of the code? Not sure if it's worth it though as we currently rely on the pagination and if the url changes we might need code changes anyway as itll work differently.

MartijnKaijser commented:
And that would need a recompiled version with changes which we never do for past releases. So if this is relying on a moving target api, hardcoding it in core is a really bad idea.

Montellese commented on system/keyboardlayouts/chinese.xml:
```diff
@@ -0,0 +1,47 @@
+Default font lacks support for all characters
+-->
+<keyboardlayouts>
+  <layout language="Chinese" layout="BasePY" codingtable="BasePY">
```
What's the `layout` attribute for? It's value is identical to the `codingtable` attribute and does not seem to be used in the code. Furthermore having a `layout` attribute in a `<layout>` tag is rather odd.

taxigps replied:
Different layout may use the same codingtable. e.g. <layout language="Chinese" layout="BasePY_QWERTY" codingtable="BasePY">

Montellese replied:
Ah sorry I just realized that the `layout` attribute is not an addition of this PR but exists in all keyboardlayouts. Nevermind my comment then.

Montellese commented:
Well like I said the ideal solution would be if every coding table would be implemented as an addon. Because then the addon could be updated whenever the API changes without any changes necessary to core.
But since that's quite a bit of work it could be put in the `<layout>` tag (as an attribute or child element). But that would require that `IInputCodingTable` would optionally support deserializing the XML layout definition.

taxigps commented:
Thanks @Montellese. Have updated as your advice. And I'll merge to one commit when all things ok. Still not have a clear idea about how to use `<layout>` to save the API messages now.

taxigps commented:
@Montellese changed to put api url in `<layout>`.

Montellese commented on xbmc/input/InputCodingTableFactory.cpp:
```diff
@@ -0,0 +1,42 @@
+  if (strTableName == "BaiduPY")
+  {
+    const char* apiurl = element->Attribute("apiurl");
+    if (apiurl == NULL)
```
Please use `nullptr` instead of `NULL`.

Montellese commented:
@taxigps: Thanks for moving the API URL into the XML.

taxigps commented:
@Paxxi @Montellese @MartijnKaijser updated and reabased. any thing else need to do?

Paxxi commented:
Nothing more required by me.

MartijnKaijser commented on xbmc/input/InputCodingTable.h:
```diff
@@ -0,0 +1,36 @@
+
+/*
+*      Copyright (C) 2005-2013 Team XBMC
+*      http://xbmc.org
```
Team Kodi   and http://kodi.tv (same for other files)

taxigps commented:
@MartijnKaijser Done.

taxigps commented:
@Paxxi who will merge this?

taxigps commented:
jenkins build this please

MartijnKaijser commented:
@taxigps @Montellese @Paxxi the switching between different keyboard is now broken. If you have selected several keyboard the button to switch them doesn't work any more

taxigps commented:
@MartijnKaijser how to reproduce the bug? I have selected `Chinese BaiduPY` and `English QWERTY`, the layout switch button works well.

MartijnKaijser commented:
seems it already works again
